presto-main/src/main/java/io/prestosql/cost/PlanNodeLocalCostEstimate.java

```diff
@@ -83,7 +83,8 @@ public PlanCostEstimate add(PlanCostEstimate costEstimate)
     {
         return new PlanCostEstimate(
                 costEstimate.getCpuCost() + this.cpuCost,
-                costEstimate.getMemoryCost() + this.maxMemory,
+                costEstimate.getMaxMemory() + this.maxMemory,
```

Comment:
what about cases:
* local operator has not started. However, some operators up the operator stream are working (e.g: join build side could have not started, yet aggregation below works)
* local operator is consuming input and source operators are outputting.

It seems we are missing case here, do we?

It seems that we should have:
`peakMemoryWhenOutputting`
`peakMemoryBeforeOutputting`

`peakMemoryAfterOutputting` is implicit (it equals 0)
`peakMemory` is also implicit (`=max(peakMemoryWhenOutputting, peakMemoryBeforeOutputting)`)

Reply:
Example of operator that uses less memory when working: e.g: join that spills
Example of operator that uses more memory when producing output: e.g: some operators that cache results in a map?

Reply:
A better example. For join node it's `local.peakMemoryBeforeOutputting` actually should take into account the whole build side (e.g: `max(build.maxMemoryBeforeOutputting, build.maxMemoryWhenOutputting)`) as join won't produce any row until build side is completed. In this case it's easy to imaging plan where join "cumulatively" (with build side only) uses more memory before producing a single row then when rows are being produced.

An example that uses (sharply) more memory when producing rows is index join. It won't consume any memory until the first probe row arrives.

This is a really neat problem!

Reply:
> Example of operator that uses less memory when working: e.g: join that spills

Our cost model currently does not take spillng into account.

Reply:
Please see second example (https://github.com/prestosql/presto/pull/247#discussion_r258162091).
* Join before outputting should include build side peak mem (which can be much larger then when join is outputting)
* Index join, partial aggregations use sharply more memory when outputting then before outputting.